pulp_container/app/registry_api.py

```diff
@@ -475,6 +479,161 @@ def get_queryset(self, *args, **kwargs):
         return (public_repositories | accessible_repositories).distinct()
 
 
+class FlatpakIndexDynamicView(APIView):
+    """
+    Handles requests to the /index/dynamic endpoint
+    """
+
+    def get(self, request):
+        unsatisfiable = False
+        req_repository = None
+        req_tags = set()
+        req_os = None
+        req_architecture = None
+        req_label_exists = {"org.flatpak.ref"}
+        req_label_values = {}
+        for key, values in request.query_params.lists():
+            if key == "repository":
+                req_repository = values[0]
+                if any(v != req_repository for v in values[1:]):
+                    unsatisfiable = True
+                    break
+            elif key == "tag":
+                req_tags.update(values)
+            elif key == "os":
+                req_os = values[0]
+                if any(v != req_os for v in values[1:]):
+                    unsatisfiable = True
+                    break
+            elif key == "architecture":
+                req_architecture = values[0]
+                if any(v != req_architecture for v in values[1:]):
+                    unsatisfiable = True
+                    break
+            elif key.startswith("label:"):
+                if key.endswith(":exists"):
+                    if any(v != "1" for v in values):
+                        # Anything other than "1" causes undefined behavior, so treat it as
+                        # unsatisfiable for simplicity:
+                        unsatisfiable = True
+                        break
+                    label = key[len("label:") : len(key) - len(":exists")]
+                    req_label_exists.add(label)
+                else:
+                    if any(v != values[0] for v in values[1:]):
+                        unsatisfiable = True
+                        break
+                    label = key[len("label:") :]
+                    req_label_values[label] = values[0]
+            else:
+                # This covers unknown query parameters (which we consider as unsatisfiable for
+                # simplicity, rather than causing some error response) as well as any annotation:...
+                # query parameters (whose presence necessarily cause the request to be
+                # unsatisfiable, as this implementation never announces any Annotations):
```

Comment:
@owtaylor PTAL

Reply:
> reading explanation from Owen i read it as you whether ignore `annotations` query param without making the request invalid or treat interchangeably annotations as labels if they are present in the request
> 
> > So we switche
> > dover to using labels for the metadata. We have the ability in our pipeline to
> > generate either or both, so you might see images with both - but the only
> > thing you need to support is labels.
> 
> can we get more clarification here?

There are two options I'd consider valid:

 * Treat requests with annotation parameters as invalid (501 or 400)
 * Fully handle annotation parameters based on the actual annotations in OCI images

I would not suggest ignoring annotation parameters or applying them to the labels.

If Pulp isn't currently indexing annotations in the database, then the first one seems easier and is perfectly sufficient for Flatpak. All support for annotations was dropped from Flatpak prior to the 1.6.0 release in Dec 2019.

Reply:
* I do not know any existing `/index/static` flatpak indexer endpoint that reports Annotations. 
* At least <https://registry.fedoraproject.org/index/static?label:org.flatpak.ref:exists=1> does not report any.  However, for an image I that it reports, the corresponding <https://registry.fedoraproject.org/v2/$base_path/manifests/$digest> is either of type `application/vnd.docker.distribution.manifest.v2+json` or `application/vnd.oci.image.manifest.v1+json`.  And if it is of the latter type, then (Observation 1:) the manifest does contain Annotations, exactly duplicating the Labels of that image I as reported by the flatpak indexer. 
* So when such an image I is imported into the existing pulp_container implementation, the Annotations in the `application/vnd.oci.image.manifest.v1+json` are available in the manifest blob, and I could in turn report them as Annotations from my pulp_container flatpak indexer endpoint.  But as those Annotations are exactly the same (due to Observation 1) as the Labels already reported by my pulp_container flatpak indexer endpoint, I see little value in doing so.
* So my pulp_container flatpak indexer endpoint never reports any Annotations (apparently in line with other existing flatpak indexer endpoints, see above).
* So for an `?annotation:...` request, my pulp_container flatpak indexer endpoint could conceivably either (1) return an empty set of responses (as no images can match), or (2) return some 400 or similar error response.  I decided for (1) because that looks like the logically most plausible option to me, but if (2) is preferred for some reason, then we could use that.  @owtaylor ?

Reply:
I would treat this as invalid request:
1) All support for annotations was dropped from Flatpak prior to the 1.6.0 release in Dec 2019. I would even mention this in docs.
2) Flatpack client itself does not send annotations in the request.

If you're not parsing manifest for annotations (I did not find any logic in your code)  you can't really be sure whether there will be any images matched or not. Today you cannot find such examples, however, you cannot go forward with this assumption since in future this can change.

Reply:
> If you're not parsing manifest for annotations (I did not find any logic in your code)

That's correct.

> you can't really be sure whether there will be any images matched or not.

That depends on how you define matching and how you define the link between  `application/vnd.oci.image.manifest.v1+json` manifest annotations and `/index/static` annotations.  (Whether or not image I should match an `?annotation:...` request when the information returned by `/index/static` about image I: (1) does not list any annotations, but (2) references an `application/vnd.oci.image.manifest.v1+json` manifests that does contain annotations matching the `?annotation:...` request.) 

> Today you cannot find such examples,

That's not correct:  Both e.g. the existing <https://registry.fedoraproject.org/index/static>, as well as my implementation here, both do reference `application/vnd.oci.image.manifest.v1+json` manifests that contain annotations, see my "Observation: 1" above.  (They just both never include those manifest's annotations in their `/index/static` JSON replies.)

>  however, you cannot go forward with this assumption since in future this can change.

see above


But anyway, I'll just change my implementation to consider any annotation requests as invalid and be done with it.

Reply:
> But anyway, I'll just change my implementation to consider any annotation requests as invalid and be done with it.

done, <https://github.com/pulp/pulp_container/pull/1301/commits/1f6b2cbe768b0c3e7871a3956385b8ab36d0d3b9> "Make annotation query parameters cause 400 response"